### ISSUE TITLE
homography_warp3d feeds a (d, x, y) grid to grid_sample, which reads (x, y, z), so warp_perspective3d fails on an identity warp

### ISSUE THREAD

**Issue by VanniLeonardo** [kornia/kornia#4502]
### What happened?

`create_meshgrid3d` emits its channels in `(z, x, y)` order, but `torch.nn.functional.grid_sample` reads a 5D grid's last dimension as `(x, y, z)`. `homography_warp3d` passes the grid straight to `grid_sample`, so the two disagree and every warp built on that path samples the wrong voxel.

The effect is not subtle and it is not an `align_corners` question. An identity homography on a 4x4x4 cube comes back with a maximum deviation of `45.0` at `align_corners=True` and `55.125` at `False`, on an input whose values run 0 to 63.

`warp_perspective3d` and `homography_warp3d` are affected. `warp_affine3d` is not, because it builds its grid with `F.affine_grid`, which already produces `(x, y, z)`.

The 2D `create_meshgrid` is correct, so this is a mismatch between the two functions rather than a library-wide convention. Feeding `grid_sample` a hand-built `(x, y, z)` grid with the same normalization reproduces the input exactly under both `align_corners` settings, which isolates the channel order as the whole of the defect.

Nothing pins this today. `warp_perspective3d`'s only numerical test asserts that an unbatched matrix gives the same answer as a batched one, and both are wrong in the same way, so it passes.

### Minimal reproduction

```python
import torch
import torch.nn.functional as F
from kornia.geometry import create_meshgrid, create_meshgrid3d
from kornia.geometry.transform import warp_perspective3d, warp_affine3d, homography_warp3d

# 1. the channel order create_meshgrid3d emits, against the 2D convention
g3 = create_meshgrid3d(2, 3, 4, normalized_coordinates=False)   # (1, D, H, W, 3)
g2 = create_meshgrid(3, 4, normalized_coordinates=False)        # (1, H, W, 2)
print("2D create_meshgrid  [h=2, w=3] ->", g2[0, 2, 3].tolist())
print("3D create_meshgrid3d[d=1, h=2, w=3] ->", g3[0, 1, 2, 3].tolist())

# 2. what that costs: an identity warp on a 4x4x4 cube
n = 4
vol = torch.arange(float(n ** 3)).view(1, 1, n, n, n)
I4 = torch.eye(4)[None]
for ac in (True, False):
    p = warp_perspective3d(vol, I4, (n, n, n), align_corners=ac)
    a = warp_affine3d(vol, I4[:, :3], (n, n, n), align_corners=ac)
    h = homography_warp3d(vol, I4, (n, n, n), align_corners=ac)
    print(f"align_corners={ac!s:5s}  warp_perspective3d {(p - vol).abs().max():9.4f}"
          f"   homography_warp3d {(h - vol).abs().max():9.4f}   warp_affine3d {(a - vol).abs().max():.3e}")

# 3. a hand-built (x, y, z) grid, same normalization, is exact
zs, ys, xs = torch.meshgrid(*[torch.arange(float(n))] * 3, indexing="ij")
for ac in (True, False):
    if ac:
        grid = torch.stack([2 * xs / (n - 1) - 1, 2 * ys / (n - 1) - 1, 2 * zs / (n - 1) - 1], -1)[None]
    else:
        grid = torch.stack([(2 * xs + 1) / n - 1, (2 * ys + 1) / n - 1, (2 * zs + 1) / n - 1], -1)[None]
    out = F.grid_sample(vol, grid, align_corners=ac)
    print(f"hand-built (x, y, z) grid, align_corners={ac!s:5s} -> {(out - vol).abs().max():.4f}")
```

Observed:

```
2D create_meshgrid  [h=2, w=3] -> [3.0, 2.0]
3D create_meshgrid3d[d=1, h=2, w=3] -> [1.0, 3.0, 2.0]     grid_sample wants [3.0, 2.0, 1.0]
align_corners=True   warp_perspective3d   45.0000   homography_warp3d   45.0000   warp_affine3d 3.815e-06
align_corners=False  warp_perspective3d   55.1250   homography_warp3d   55.1250   warp_affine3d 0.000e+00
hand-built (x, y, z) grid, align_corners=True  -> 0.0000
hand-built (x, y, z) grid, align_corners=False -> 0.0000
```

Expected: an identity homography reproduces the input, and `warp_perspective3d` agrees with `warp_affine3d` given the same identity, exactly as the 2D pair do.

### Environment

- Kornia version: 0.9.0rc1 (current `main`)
- PyTorch version: 2.9.1+cu128
- Python version: 3.11.14
- OS / device: Linux, CPU

### Anything else?

Whichever side is changed, the fix needs a sweep of the `create_meshgrid3d` callers, since anything that consumes the grid as `(z, x, y)` today would have to move with it.

This is separate from the `align_corners` defect in the 3D normalization, which I am filing alongside it. That one affects `warp_affine3d`; this one does not.

Found while verifying the 3D follow-up to #3904 that came out of the #3945 review. Investigated with Claude Code; every figure above is from an executed run on current `main`.

**Comment by VanniLeonardo** [comment]
Correcting my own framing before anyone acts on it. The defect is real and the numbers stand, but I blamed the wrong function.

`create_meshgrid3d`'s `(d, x, y)` order is not a bug. It is a documented kornia convention, and two other places depend on it.

`normalize_pixel_coordinates3d` states it outright:

> `pixel_coordinates` is `(*, 3)` in `(d, x, y)` order, depth first, then x, then y, *not* `(x, y, z)`. This is the order `create_meshgrid3d` produces.

`conv_soft_argmax3d` returns coordinates in that same order, and its public output would silently change if the grid changed. On a single hot voxel at depth 2, x 4, y 1:

```
conv_soft_argmax3d coords -> [2.0, 4.0, 1.0]      # (d, x, y), matching its own comments
                                                   # (x, y, z) would have been [4.0, 1.0, 2.0]
```

`_distance_transform_3d_impl` is the third caller and is immune, since it takes a norm over the last axis of a symmetric cube.

So the contract is broken at the call site, not in the grid. `homography_warp3d` hands its `(d, x, y)` grid to `warp_grid3d`, whose own Convention block says it takes `(x, y, z)`, and then to `grid_sample`, which reads a 5D grid the same way. Neither the grid nor the consumers are wrong on their own terms. The 3D warp path simply never converts between the two.

That makes the fix local and much smaller than my "sweep of the callers" line suggested. Reordering to `(x, y, z)` at that one call site is enough:

```python
grid = create_meshgrid3d(depth, height, width, normalized_coordinates=..., device=...)
grid = grid[..., [1, 2, 0]]   # kornia's (d, x, y) into grid_sample's (x, y, z)
```

Measured with only that change, an identity warp becomes exact, and the residual is precisely the separate `align_corners` defect in #4503:

| dsize | `align_corners=True` | `align_corners=False` |
|---|---|---|
| (4, 4, 4) | 0.0 | 55.125 |
| (2, 3, 4) | 0.0 | 20.125 |
| (1, 5, 3) | 0.0 | 10.5 |

So the two issues separate cleanly. This one is the channel order and accounts for the whole `align_corners=True` error. #4503 is the normalization convention and accounts for the whole remaining `False` error. Fixing this one first is also what makes #4503 verifiable through `warp_perspective3d` rather than only through `warp_affine3d`.

The title is now wrong too. A better one is "homography_warp3d feeds a (d, x, y) grid to grid_sample, which reads (x, y, z), so warp_perspective3d fails on an identity warp".

I am taking this one unless you would rather it went to someone else.
